datafusion/physical-plan/src/sorts/merge.rs

```diff
@@ -156,12 +164,22 @@ impl<C: CursorValues> SortPreservingMergeStream<C> {
         }
         // try to initialize the loser tree
         if self.loser_tree.is_empty() {
-            // Ensure all non-exhausted streams have a cursor from which
-            // rows can be pulled
-            for i in 0..self.streams.partitions() {
-                if let Err(e) = ready!(self.maybe_poll_stream(cx, i)) {
```

Comment:
It looks to me like the issue is that the `ready!` macro `return`s if the poll isn't ready -- and thus the other streams aren't polled. 

I wonder if simply changing this to not use `ready!` would work, and be a smaller change?

Like could you change this to be the following instead? I think that would ensure each stream is polled

```rust
                if let Poll::Ready(Err(e)) = ready!(self.maybe_poll_stream(cx, i)) {

Reply:
I guess you meant
`if let Poll::Ready(Err(e)) = self.maybe_poll_stream(cx, i)`
This change solves my problem, but some existing tests fails:
```
thread 'sorts::sort_preserving_merge::tests::test_async' panicked at datafusion/physical-plan/src/sorts/sort_preserving_merge.rs:959:9:
assertion `left == right` failed
  left: 0
 right: 1
```

```
assertion failed: poll.is_pending()
thread 'sorts::sort_preserving_merge::tests::test_drop_cancel' panicked at datafusion/physical-plan/src/test.rs:42:5:
assertion failed: poll.is_pending()
```

I suppose that's because of we do not return pending in case of any of the partitions returns pending (what I understand is if there is at least one pending partition, SPM should return pending).

Reply:
I coded up what I had in mind here: https://github.com/synnada-ai/datafusion-upstream/pull/34